drivers/can/can_mcux_flexcan_fd.c

```diff
@@ -0,0 +1,1091 @@
+
+	FLEXCAN_EnterFreezeMode(config->base);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int mcux_flexcan_fd_set_mode(const struct device *dev, can_mode_t mode)
+{
+	const struct mcux_flexcan_fd_config *config = dev->config;
+	struct mcux_flexcan_fd_data *data = dev->data;
+	uint32_t ctrl1;
+	uint32_t mcr;
+	uint32_t fdctrl;
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+
+	if ((mode &
+	     ~(CAN_MODE_FD | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_3_SAMPLES)) != 0) {
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	ctrl1 = config->base->CTRL1;
+	mcr = config->base->MCR;
+	fdctrl = config->base->FDCTRL;
+
+	if ((mode & CAN_MODE_LOOPBACK) != 0) {
+		/* Enable loopback and self-reception */
+		ctrl1 |= CAN_CTRL1_LPB_MASK;
+		mcr &= ~(CAN_MCR_SRXDIS_MASK);
+		/* TDC must be disabled when Loop Back mode is enabled */
+		fdctrl &= ~(CAN_FDCTRL_TDCEN_MASK);
+	} else {
+		/* Disable loopback and self-reception */
+		ctrl1 &= ~(CAN_CTRL1_LPB_MASK);
+		mcr |= CAN_MCR_SRXDIS_MASK;
+		/* TDC is enabled */
+		fdctrl |= CAN_FDCTRL_TDCEN_MASK;
+	}
+
+	if ((mode & CAN_MODE_LISTENONLY) != 0) {
+		/* Enable listen-only mode */
+		ctrl1 |= CAN_CTRL1_LOM_MASK;
+	} else {
+		/* Disable listen-only mode */
+		ctrl1 &= ~(CAN_CTRL1_LOM_MASK);
+	}
+
+	if ((mode & CAN_MODE_3_SAMPLES) != 0) {
+		/* Enable triple sampling mode */
+		ctrl1 |= CAN_CTRL1_SMP_MASK;
+	} else {
+		/* Disable triple sampling mode */
+		ctrl1 &= ~(CAN_CTRL1_SMP_MASK);
+	}
+
+	config->base->CTRL1 = ctrl1;
+	config->base->MCR = mcr;
+	config->base->FDCTRL = fdctrl;
+	return 0;
+}
+
+static void mcux_flexcan_fd_from_can_frame(const struct can_frame *src,
+					   flexcan_fd_frame_t *dest)
+{
+	memset(dest, 0, sizeof(*dest));
+
+	if ((src->flags & CAN_FRAME_IDE) != 0) {
+		dest->format = kFLEXCAN_FrameFormatExtend;
+		dest->id = FLEXCAN_ID_EXT(src->id);
+	} else {
+		dest->format = kFLEXCAN_FrameFormatStandard;
+		dest->id = FLEXCAN_ID_STD(src->id);
+	}
+
+	if ((src->flags & CAN_FRAME_RTR) != 0) {
+		dest->type = kFLEXCAN_FrameTypeRemote;
+	} else {
+		dest->type = kFLEXCAN_FrameTypeData;
+	}
+
+	dest->length = src->dlc;
+
+	if ((CAN_FRAME_BRS & src->flags) != 0) {
+		dest->brs = 1U;
+	}
+	for (uint8_t index = 0U; index <= CANFD_MAX_DLC; index++) {
+		dest->dataWord[index] = sys_cpu_to_be32(src->data_32[index]);
+	}
+}
+
+static void mcux_flexcan_fd_to_can_frame(const flexcan_fd_frame_t *src,
+					 struct can_frame *dest)
+{
+	memset(dest, 0, sizeof(*dest));
+
+	if (src->format == kFLEXCAN_FrameFormatStandard) {
+		dest->id = FLEXCAN_ID_TO_CAN_ID_STD(src->id);
+	} else {
+		dest->flags |= CAN_FRAME_IDE;
+		dest->id = FLEXCAN_ID_TO_CAN_ID_EXT(src->id);
+	}
+
+	if (src->type == kFLEXCAN_FrameTypeRemote) {
+		dest->flags |= CAN_FRAME_RTR;
+	}
+
+	dest->dlc = src->length;
+	dest->flags |= CAN_FRAME_FDF;
```

Comment:
You cannot unconditionally set the `CAN_FRAME_FDF` bit. The driver should work in both classic and CAN-FD modes.

Reply:
[Now](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/drivers/can/can_mcux_flexcan_fd.c#L270) the received flags are set and the can frame type flags are set correspondingly.